### DISCUSSION TITLE
go2rtc: first impressions

### DISCUSSION THREAD

**Discussion by banthungprong** [blakeblackshear/frigate#4235] [General]
After seeing a lot of activity regarding commits I wanted to try these immediately.
So updated my fork and started it in VS Code.

Of course my biggest interest is the new go2rtc integration. This is phantastic!

After changing the config from rtmp to restream it worked right away.
In the live view switching the player is easy and works (jsmpeg, mse and webrtc).

I now use webrtc in the live view - learned to love it when streaming with Janus before.
The configuration of the necessary yaml-File is easy and already documented.

Now I get Full-HD streams from my HikVision-Cams in Frigate. Perfect quality.
And even better: 4K from my noname cam with Audio is shown in the live view.

Some restrictions/observations:
- H265/H265+ doesn't work, I had to switch the cams from H265 to H264. From the go2rtc-site H265 should be possible. Will look further why it isn't working.
- 4K stream and audio seems a bit slow
- the browser console shows some errors (DOMException etc).

Now I will build a docker image and use this one for more testing of all the new features.
I want to thank Nick and Blake for the excellent new feature.

**Comment by NickM-27** [comment]
> H265/H265+ doesn't work, I had to switch the cams from H265 to H264. From the go2rtc-site H265 should be possible. Will look further why it isn't working.

See https://github.com/AlexxIT/go2rtc#codecs-madness : only Safari supports webRTC with H.265. MSE H.265 requires browser support which must be enabled.

> 4K stream and audio seems a bit slow

With which streaming protocol? (MSE or webRTC)

> the browser console shows some errors (DOMException etc).

These are expected errors as the RTCPeerConnection tries different candidates for a connection, normal and nothing to worry about.

**Reply by banthungprong** [reply]
I use a beta of chrome/chromium (109.0....) with H265 support - could view the H265 recordings before already.
webrtc for 4k
I know the errors like this from Janus (meetecho). Sometimes there are some relevant errors in the sea of normal errors ;-)

**Reply by NickM-27** [reply]
> I use a beta of chrome/chromium (109.0....) with H265 support - could view the H265 recordings before already.

Right, webRTC and playback for h.265 is different. If MSE h.265 isn't working then seeing the error would be helpful. 

> webrtc for 4k

Interesting, 5 MP working fine here. Will be curious to see if others see this.

**Comment by banthungprong** [comment]
H265 with MSE isn't working. Error will follow soon. Right now I use a 4k stream with reduced bitrate in H264 and cast this to my big TV. It feels a little bit like being there in Thailand.

I'm using a docker image (banac/frigate:latest) on my GK41 now. 

Again: Thanks a lot. Great solution offered here.

**Reply by banthungprong** [reply]
H265 Stream, Full-HD:
mse:
frigate  | 17:02:04.189 WRN [api.mse] init error="empty SPS: &streamer.Codec{Name:\"H265\", ClockRate:0x15f90, Channels:0x0, FmtpLine:\"\", PayloadType:0x60}"

webrtc:
frigate  | 17:02:20.879 ERR [api.ws] readJSON error="websocket: close 1005 (no status)"

**Reply by NickM-27** [reply]
> frigate | 17:02:20.879 ERR [api.ws] readJSON error="websocket: close 1005 (no status)"

That's an unrelated error that will happen every time you change from a stream type since frigate is closing the websocket to go2rtc. Even if you open a camera's webRTC and change to a different camera you will see that every time.

> H265 Stream, Full-HD:
> mse:
> frigate | 17:02:04.189 WRN [api.mse] init error="empty SPS: &streamer.Codec{Name:"H265", ClockRate:0x15f90, Channels:0x0, FmtpLine:"", PayloadType:0x60}"

thanks will see if this is expected or not

**Reply by AlexxIT** [reply]
Empty SPS can be fixed. It's better to create related issues in go2rtc project. So I can know about them :)

**Reply by banthungprong** [reply]
Fix will be available in the next release.
See
https://github.com/AlexxIT/go2rtc/issues/99#issuecomment-1326233814

**Comment by spacebares** [comment]
Havin trouble with reolink audio live using webRTC. There is audio in the events just not live. MSE same problem.
You have anything special in ffmpeg for live audio ?

**Reply by NickM-27** [reply]
webRTC requires audio to be opus or pcma / pcmu. MSE requires AAC audio.

You can use the option:

```yaml
restream:
  force_audio: true
```

which will force the audio to be opus and compatible with WebRTC

**Reply by NickM-27** [reply]
https://github.com/AlexxIT/go2rtc#codecs-madness

**Reply by spacebares** [reply]
Nice! That was it.

**Reply by hagak** [reply]
I have been messing with go2rtc for a bit, actually gave up because I just could not make it reliable for more then a single camera at a time.  If I had a HA lovelace card with 4 cameras eventually one or more would error out. details if curious: https://github.com/AlexxIT/go2rtc/issues/90#issuecomment-1295245407

However this topic "restream" was mentioned twice but I can not find much info on it here or on go2rtc.  Is this a go2rtc configuration?

**Reply by NickM-27** [reply]
> However this topic "restream" was mentioned twice but I can not find much info on it here or on go2rtc. Is this a go2rtc configuration?

No, it is a frigate configuration, as go2rtc will be built in to frigate in 0.12 and the RTMP based restream will be considered deprecated in favor of the new RTSP restream (which is using go2rtc behind the scenes)

**Reply by hagak** [reply]
So that probably wont solve my issue.

**Reply by NickM-27** [reply]
> So that probably wont solve my issue.

I mean, your issue seems to be with the RTMP restream, which we are going to recommend not be used in 0.12. Also, I have been using this `dev` version with go2rtc integrated for over a month now and have had no issues viewing streams including multiple streams at the same time.

**Reply by hagak** [reply]
Sorry i confused with the terminology.  I do not have any restream: configuration in frigate.  Heck I can not even find reference to it on the docs page for the configuration. https://docs.frigate.video/configuration/index#full-configuration-reference

If I am using it how do I go about not using it?

**Reply by NickM-27** [reply]
> Sorry i confused with the terminology.  I do not have any restream: configuration in frigate.  Heck I can not even find reference to it on the docs page for the configuration. https://docs.frigate.video/configuration/index#full-configuration-reference
> 
> 
> 
> If I am using it how do I go about not using it?

I am using restream right now as a generic term, as in taking the cameras stream and making it available to other software to read. Your issue in go2rtc is (as I understand) using RTMP in frigate (which is the current method for restreaming a camera feed). 

In 0.12 we are going to recommend users NOT use RTMP and it will be removed in a future release. Instead, a new restream method (just called restream in 0.12) will be available and recommended and will use go2rtc behind the scenes.

**Reply by spacebares** [reply]
> So that probably wont solve my issue.

check this out here: https://github.com/dermotduffy/frigate-hass-card#android-will-not-render-4-jsmpeg-live-views might be worth a try.

**Comment by banthungprong** [comment]
Just made a new image based on the latest commits.
Everything is working. 
I like the new debug view with more details for each camera and the ffprobe integration. Very nice.

![image](https://user-images.githubusercontent.com/59219161/202837484-03f4248c-a24b-4830-b273-b0f7abf0a573.png)

**Reply by NickM-27** [reply]
Just so you know, frigate now has a build done on every commit to dev. It's in the packages section on the right side

**Reply by banthungprong** [reply]
Thanks. That's very nice. Will use it when I get lazy. But I like to build it myself and test it. Just having fun.

**Comment by Cold-Lemonade** [comment]
@NickM-27 I'd like to give the dev build a try. How do I install it? I assume I need to do something with this URL: https://github.com/blakeblackshear/frigate.git.

**Reply by NickM-27** [reply]
We have prebuilt images that are done on each dev commit. Just check the packages section of the repo.

**Reply by Cold-Lemonade** [reply]
Many thanks!
